packages/beacon-node/src/sync/utils/downloadByRoot.ts

```diff
@@ -0,0 +1,589 @@
+        peer: peerIdStr,
+      });
+    }
+    for (const columnSidecar of columnSidecars) {
+      blockInput.addColumn({
+        columnSidecar,
+        blockRootHex: rootHex,
+        seenTimestampSec: Date.now(),
+        source: BlockInputSource.byRoot,
+        peerIdStr,
+      });
+    }
+  }
+
+  let status: PendingBlockInputStatus;
+  let timeSyncedSec: number | undefined;
+  if (blockInput.hasBlockAndAllData()) {
+    status = PendingBlockInputStatus.downloaded;
+    timeSyncedSec = Date.now() / 1000;
+  } else {
+    status = PendingBlockInputStatus.pending;
+  }
+
+  return {
+    status,
+    blockInput,
+    timeSyncedSec,
+    timeAddedSec: cacheItem.timeAddedSec,
+    peerIdStrings: cacheItem.peerIdStrings,
+  };
+}
+
+export async function fetchByRoot({
+  config,
+  network,
+  executionEngine,
+  peerIdStr,
+  blockRoot,
+  cacheItem,
+}: FetchByRootProps): Promise<FetchByRootResponses> {
+  let block: SignedBeaconBlock;
+  let blobSidecars: deneb.BlobSidecars | undefined;
+  let columnSidecars: fulu.DataColumnSidecars | undefined;
+
+  if (isPendingBlockInput(cacheItem)) {
+    if (cacheItem.blockInput.hasBlock()) {
+      block = cacheItem.blockInput.getBlock();
+    } else {
+      block = await fetchAndValidateBlock({
+        config,
+        network,
+        peerIdStr,
+        blockRoot,
+      });
+    }
+
+    const forkName = config.getForkName(block.message.slot);
+    if (!cacheItem.blockInput.hasAllData()) {
+      if (isBlockInputBlobs(cacheItem.blockInput)) {
+        blobSidecars = await fetchAndValidateBlobs({
+          config,
+          network,
+          executionEngine,
+          peerIdStr,
+          forkName: forkName as ForkPreFulu,
+          block: block as SignedBeaconBlock<ForkPostDeneb>,
+          blockRoot,
+          blobMeta: cacheItem.blockInput.getMissingBlobMeta(),
+        });
+      }
+      if (isBlockInputColumns(cacheItem.blockInput)) {
+        columnSidecars = await fetchAndValidateColumns({
+          config,
+          network,
+          executionEngine,
+          peerIdStr,
+          forkName: forkName as ForkPostFulu,
+          block: block as SignedBeaconBlock<ForkPostFulu>,
+          blockRoot,
+          columnMeta: cacheItem.blockInput.getMissingSampledColumnMeta(),
+        });
+      }
+    }
+  } else {
+    block = await fetchAndValidateBlock({
+      config,
+      network,
+      peerIdStr,
+      blockRoot,
+    });
+    const forkName = config.getForkName(block.message.slot);
+    if (isForkPostFulu(forkName)) {
+      columnSidecars = await fetchAndValidateColumns({
+        config,
+        network,
+        executionEngine,
+        peerIdStr,
+        forkName,
+        blockRoot,
+        block: block as SignedBeaconBlock<ForkPostFulu>,
+        columnMeta: {
+          missing: network.custodyConfig.sampledColumns,
+          versionedHashes: (block as SignedBeaconBlock<ForkPostFulu>).message.body.blobKzgCommitments.map((c) =>
```

Comment:
in #8282 we don't call `getBlobsV2()` in this flow so no need `versionedHashed`
also it's a performance concern here that we call `digest()` repetitively in this flow via kzgCommitmentToVersionedHash()` util